src/wallet/rpc/wallet.cpp

```diff
@@ -582,6 +582,124 @@ static RPCHelpMan upgradewallet()
+{
+    std::shared_ptr<const CWallet> const wallet = GetWalletForJSONRPCRequest(request);
+    if (!wallet) return NullUniValue;
+    const CWallet* pwallet = wallet.get();
+
+    RPCTypeCheck(request.params, {UniValue::VARR, UniValue::VOBJ}, true);
+
+    LOCK(pwallet->cs_wallet);
+
+    CAmount changes{0};
+    std::set<COutPoint> spent;
+
+    UniValue include_watchonly(UniValue::VNULL);
+    if (request.params[1].isObject()) {
+        UniValue options = request.params[1];
+        RPCTypeCheckObj(options,
+            {
+                {"include_watchonly", UniValueType(UniValue::VBOOL)},
+            },
+            true, true);
+
+        include_watchonly = options["include_watchonly"];
+    }
+
+    isminefilter filter = ISMINE_SPENDABLE;
+    if (ParseIncludeWatchonly(include_watchonly, *pwallet)) {
+        filter |= ISMINE_WATCH_ONLY;
+    }
+
+    const auto& txs = request.params[0].get_array();
+    std::map<COutPoint, CAmount> new_utxos; // UTXO:s that were made available in transaction array
+    for (size_t i = 0; i < txs.size(); ++i) {
+        CMutableTransaction mtx;
+        if (!DecodeHexTx(mtx, txs[i].get_str(), /* try_no_witness */ true, /* try_witness */ true)) {
+            throw JSONRPCError(RPC_DESERIALIZATION_ERROR, "Transaction hex string decoding failure.");
+        }
+
+        // Fetch previous transactions (inputs)
+        std::map<COutPoint, Coin> coins;
+        for (const CTxIn& txin : mtx.vin) {
+            coins[txin.prevout]; // Create empty map entry keyed by prevout.
+        }
+        pwallet->chain().findCoins(coins);
+
+        // Fetch debit; we are *spending* these; if the transaction is signed and
+        // broadcast, we will lose everything in these
+        for (const auto& txin : mtx.vin) {
+            auto& op = txin.prevout;
+            if (spent.count(op)) {
+                throw JSONRPCError(RPC_INVALID_PARAMETER, "Transaction(s) are spending the same output more than once");
+            }
+            if (new_utxos.count(op)) {
+                changes -= new_utxos.at(op);
+                new_utxos.erase(op);
+            } else {
+                if (!coins.count(op)) {
+                    throw JSONRPCError(RPC_INVALID_PARAMETER, "One or more transaction inputs are missing");
+                }
+                if (coins.at(txin.prevout).IsSpent()) {
+                    throw JSONRPCError(RPC_INVALID_PARAMETER, "One or more transaction inputs have been spent already");
+                }
+                changes -= pwallet->GetDebit(txin, filter);
+            }
+            spent.insert(txin.prevout);
+        }
+
+        for (size_t i = 0; i < txs.size(); ++i) {
+            // Populate new outputs
+            const auto hash = mtx.GetHash();
+            for (size_t i = 0; i < mtx.vout.size(); ++i) {
+                new_utxos[COutPoint(hash, i)] = (pwallet->IsMine(mtx.vout[i]) & filter) ? mtx.vout[i].nValue : 0;
```

Comment:
In 84a745eed211b597a89f3bbc3da73b0ac2cb457d "rpc/wallet: add simulaterawtransaction RPC"

This can be moved into the loop below that iterates the outputs.

Reply:
Moved.